File: ch01/python/ollama-local-streaming/ollama-local-streaming.py
"""
(C) Copyright 2026 Boni Garcia (https://bonigarcia.github.io/)
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
import json
import os
import time

import requests

OLLAMA_HOST = os.getenv("OLLAMA_HOST", "http://localhost:11434")
OLLAMA_MODEL = os.getenv("OLLAMA_MODEL", "gemma3:4b")


def query_model(prompt: str,
                model: str = OLLAMA_MODEL,
                temperature: float = 0) -> str:
    """Stream the response of a local LLM running through Ollama."""
    payload = {
        "model": model,
        "prompt": prompt,
        "stream": True,
        "options": {
            "temperature": temperature,
        },
    }

    start = time.perf_counter()
    first_token = None
    chunks = []
    stats = {}
    with requests.post(f"{OLLAMA_HOST}/api/generate",
                       json=payload,
                       stream=True,
                       timeout=120) as response:
        response.raise_for_status()
        for line in response.iter_lines():
            if not line:
                continue
            chunk = json.loads(line)
            text = chunk.get("response", "")
            if text and first_token is None:
                first_token = time.perf_counter() - start
            print(text, end="", flush=True)
            chunks.append(text)
            if chunk.get("done"):
                stats = chunk
    latency = time.perf_counter() - start
    print()

    input_tokens = stats.get("prompt_eval_count", 0)
    output_tokens = stats.get("eval_count", 0)
    ttft = first_token if first_token is not None else latency
    print(f"\tModel: {stats.get('model', model)}")
    print(f"\tTime to first token: {ttft:.3f} seconds")
    print(f"\tLatency: {latency:.3f} seconds")
    print(f"\tInput tokens: {input_tokens}")
    print(f"\tOutput tokens: {output_tokens}")
    print(f"\tTotal tokens: {input_tokens + output_tokens}")

    return "".join(chunks).strip()


if __name__ == "__main__":
    prompt = "How many tokens are in your context window?"

    print("User:", prompt)
    print("Local LLM: ", end="", flush=True)
    query_model(prompt)

File: ch01/javascript/ollama-local-streaming/ollama-local-streaming.js
/*
(C) Copyright 2026 Boni Garcia (https://bonigarcia.github.io/)
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { performance } from 'node:perf_hooks';

const ollamaHost = process.env.OLLAMA_HOST || 'http://localhost:11434';
const ollamaModel = process.env.OLLAMA_MODEL || 'gemma3:4b';

async function queryModel(userPrompt, model = ollamaModel, temperature = 0) {
    const payload = {
        model: model,
        prompt: userPrompt,
        stream: true,
        options: {
            temperature: temperature,
        },
    };

    const start = performance.now();
    const response = await fetch(`${ollamaHost}/api/generate`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
    });

    if (!response.ok) {
        throw new Error(`Ollama request failed: ${await response.text()}`);
    }

    const decoder = new TextDecoder();
    let firstToken = null;
    let buffer = '';
    let answer = '';
    let stats = {};

    for await (const bytes of response.body) {
        buffer += decoder.decode(bytes, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop();
        for (const line of lines) {
            if (!line.trim()) {
                continue;
            }
            const chunk = JSON.parse(line);
            const text = chunk.response || '';
            if (text && firstToken === null) {
                firstToken = (performance.now() - start) / 1000;
            }
            process.stdout.write(text);
            answer += text;
            if (chunk.done) {
                stats = chunk;
            }
        }
    }
    const latency = (performance.now() - start) / 1000;
    console.log();

    const inputTokens = stats.prompt_eval_count || 0;
    const outputTokens = stats.eval_count || 0;
    const ttft = firstToken === null ? latency : firstToken;

    console.log(`\tModel: ${stats.model || model}`);
    console.log(`\tTime to first token: ${ttft.toFixed(3)} seconds`);
    console.log(`\tLatency: ${latency.toFixed(3)} seconds`);
    console.log(`\tInput tokens: ${inputTokens}`);
    console.log(`\tOutput tokens: ${outputTokens}`);
    console.log(`\tTotal tokens: ${inputTokens + outputTokens}`);

    return answer.trim();
}

const userPrompt = 'How many tokens are in your context window?';

console.log('User:', userPrompt);
process.stdout.write('Local LLM: ');
await queryModel(userPrompt);
